completion_aggregator/aggregator.py

```diff
@@ -0,0 +1,173 @@
+"""
+Aggregator service.
+
+This service periodically determines which stale_blocks need updating, and
+enqueues tasks to perform those updates.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import collections
+import logging
+import time
+from datetime import timedelta
+
+import six
+
+from django.db import transaction
+from django.utils import timezone
+
+from . import models, tasks, utils
+
+log = logging.getLogger(__name__)
+
+
+class BasePeriodicService(object):
+    """
+    Enqueue a new service every `period` for `runtime`, then quit.
+
+    `period` and `runtime` should be provided as timedelta objects, or None.  A
+    `period` of None indicates that the task should rerun immediately upon
+    completion.
+
+    Rather than starting this service from a cronjob every `period`, we provide
+    this service, which will iterate over to amortize the cost of firing up a
+    new edx-platform management task, which itself would take longer than a
+    reasonable period.
+    """
+
+    DEFAULT_PERIOD = timedelta(seconds=0)
+
+    service_name = "Unnamed Service"
+
+    def __init__(self, runtime=None, period=None):
+        """
+        Create a PeriodicService.
+
+        **Arguments:**
+
+        runtime (timedelta|None):
+            Run the service for this length of time, then quit.  If not
+            provided, run forever.
+
+        period (timedelta|None):
+            Perform the task this frequently.  If not provided, start
+            it again as soon as it is completed.
+
+        """
+        log.debug("initializing service")
+        if period is None:
+            self.period = self.DEFAULT_PERIOD
+        else:
+            self.period = period
+        if runtime is None:
+            # Run forever (or 999 years)
+            self.runtime = timedelta(days=365 * 999)
+        else:
+            self.runtime = runtime
+
+    def start(self):
+        """
+        Start the service.
+        """
+        service_start_time = timezone.now()
+        service_end_time = service_start_time + self.runtime
+        tick = service_start_time
+        i = 0
+        while tick < service_end_time:
+            i += 1
+            if i % 1000 == 0:
```

Comment:
Seems arbitrary to run cleanup every 1000 ticks?

Reply:
Completely arbitrary.  But then, so is every other number. :man_shrugging:.

I like your idea of just running a second, slower task.